cli/command/service/list.go

```diff
@@ -73,30 +71,38 @@ func runList(dockerCli *command.DockerCli, opts listOptions) error {
 			return err
 		}
 
-		PrintNotQuiet(out, services, nodes, tasks)
-	} else if !opts.quiet {
-		// no services and not quiet, print only one line with columns ID, NAME, MODE, REPLICAS...
-		PrintNotQuiet(out, services, []swarm.Node{}, []swarm.Task{})
-	} else {
-		PrintQuiet(out, services)
+		mode, replicas = GetServicesStatus(services, nodes, tasks)
+	}
+
+	format := opts.format
+	if len(format) == 0 {
+		if len(dockerCli.ConfigFile().ServicesFormat) > 0 && !opts.quiet {
+			format = dockerCli.ConfigFile().ServicesFormat
+		} else {
+			format = formatter.TableFormatKey
+		}
 	}
 
-	return nil
+	servicesCtx := formatter.Context{
+		Output: dockerCli.Out(),
+		Format: formatter.NewServiceListFormat(format, opts.quiet),
+		Trunc:  !opts.noTrunc,
+	}
+	return formatter.ServiceListWrite(servicesCtx, services, mode, replicas)
 }
 
-// PrintNotQuiet shows service list in a non-quiet way.
-// Besides this, command `docker stack services xxx` will call this, too.
-func PrintNotQuiet(out io.Writer, services []swarm.Service, nodes []swarm.Node, tasks []swarm.Task) {
+// GetServicesStatus returns a map of mode and replicas
+func GetServicesStatus(services []swarm.Service, nodes []swarm.Node, tasks []swarm.Task) (map[string]string, map[string]string) {
```

Comment:
Should/Could this return a map of struct instead, with mode and replica in the struct, instead of two maps ? 👼